### PR TITLE
[StructuralMechanicsApplication] Rve periodic

### PR DESCRIPTION
**Pull request by RiccardoRossi**
This branch provides the tools needed (an utility and an AnalysisStage) for the evaluation of the Elastic matrix corresponding to a given RVE.

The limiation is that the geoemtry of the RVE must be periodic, and it must be delimited by a parallelepiped

### CODE REVIEW

**Comment by philbucher** [comment]
wow this is quite messy ... you sure you want to commit like this?
Asking first before I review in detail

**Comment by RiccardoRossi** [comment]
Hi @philbucher i cleaned up some (purely cosmetic) but from my point of view this is not too bad :-D

i have two main things to discuss which is the reason of being of this PR

1 - For the input, i am not sure of wherein the settings i should put the parameters to be passed to my analysis
2 - in order to have a test, i need my analysis stage to return something (the estimated matrix). We did not decide yet how this should be done. I could write to a file but it is not too clean, so i think this is something for discussion (however this PR provides a nice feature which i would love to have in for release, so if the good solution of this requires a lot of changes, i would do that in a second step)

File: applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
+from __future__ import print_function, absolute_import, division #makes KratosMultiphysics backward compatible with python 2.6 and 2.7
+
+import KratosMultiphysics
+import KratosMultiphysics.StructuralMechanicsApplication
+import math
+from structural_mechanics_analysis import StructuralMechanicsAnalysis
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py]:
```suggestion
from KratosMultiphysics.StructuralMechanicsApplication.structural_mechanics_analysis import StructuralMechanicsAnalysis
```

**Comment by philbucher** [comment]
my main three points of critisism are that you should 
1. separate header-source
2. add a test (Otherwise no guarantees can be made ;) )
3. Still there are some debugging statements

Why would you return a value from the AnalysisStage? I thought this should be the main instance in a simulation?

**Comment by RiccardoRossi** [comment]
regarding 1, ok.

regarding point 2 and 3 they go togheter: the goal of this analysis stage is to compute the ElasticityTensor which corresponds to the rve. 
As of now this is written to a file.

i was wondering if in the test i should write and read this file, or if i shall output it somehow. 

also i am not sure of where i should put the input parameters that belong to the "analysisstage"

**Comment by philbucher** [comment]
So you would call this custom stage from the regular StructuralMechanicsAnalysis?

File: applications/StructuralMechanicsApplication/tests/rve_test/test_rve_analytic.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    def test_isotropic_rve(self):
+        with open("ProjectParameters.json",'r') as parameter_file:
+            parameters = KratosMultiphysics.Parameters(parameter_file.read())
+            
+        boundary_mp_name = "Structure.SurfacePressure3D_Pressure_on_surfaces_Auto2"
+        averaging_mp_name = "Structure.computing_domain"
+        
+        model = KratosMultiphysics.Model()
+        simulation = RVEAnalysis.RVEAnalysis(model,parameters,boundary_mp_name,averaging_mp_name)
+        simulation.Run()
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/rve_test/test_rve_analytic.py]: @loumalouomega this is the test that uncovers a memory bug (as well as a bug in the results) detected by valgrind. As of now it is skipped but it breaks deterministically 

@adityaghantasala most interestingly with the block builder and solver i get the exact prediction for this testcase. With this i mean that the AVERAGE stress is ok, nevertheless the solution in terms of displacements is not the one i would expect. This is actually a very strange result since it appears to be against the uniqueness of solution...
The point is that the expected result is that 

        displacement(nodeI) = strain_tensor * InitialPosition(nodeI)

this is periodic (with jump) and complies with the boundary conditions i am applying. Since the strain field is constant, stress is constant and like in a patch test FEM should deliver exact results.

From this analysis i get in average the correct stress, nevertheless the displacement field is not linear and hence the stress field is not constant. Periodicity condition IS respected

**Comment by RiccardoRossi** [comment]
@philbucher i think i addressed all of your concerns (and indeed code is much more polished than it used to be)

I am not sure of what you are asking about the StructuralAnalysisStage.

The RVEAnalysis is derived from that ... but it performs a completely different operation (Actually it applies 6 different periodicity conditons and uses that to estimate the ELASTICITY_TENSOR of an RVE).
This however also implies that one needs to construct explicitly from the main the RVEAnalysis.
Incidentally this highlights a limitation of the current mecanism:

I believe that our project parameters should eventually have something like

       stages_list =[
           {
                "stage_type" : StructuralAnalysisStage,
                "stage_parameters" : ...
           },
           {
                "stage_type" : RVEAnalysisStage,
                "stage_parameters" : ...
           }

of course this is not to be addressed in this PR, i am just telling it since if we already had this mechanism than i would be able to use the same Main as always

**Comment by RiccardoRossi** [comment]
@loumalouomega the new test is in a subdirectory. i am not clear on how to include it in the test suite...could you guide me?

**Comment by loumalouomega** [comment]
> @loumalouomega the new test is in a subdirectory. i am not clear on how to include it in the test suite...could you guide me?

I can, let me go to your branch

**Comment by loumalouomega** [comment]
@RiccardoRossi I added the test to the suite, I will continue debugging

**Comment by RiccardoRossi** [comment]
@philbucher i get a warning of "old format" been used, but i cannot see what is wrong.

would u mind taking a look?

**Comment by philbucher** [comment]
> @philbucher i get a warning of "old format" been used, but i cannot see what is wrong.
> 
> would u mind taking a look?

@RiccardoRossi to me your input looks correct/up-to-date
can you please post the entire warning?

**Comment by RiccardoRossi** [comment]
here goes the output

~~~
 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ___ |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 6.0.0-693bc1d484-FullDebug
Importing  KratosStructuralMechanicsApplication 
    KRATOS   ___|  |                   |                   |
           \___ \  __|  __| |   |  __| __| |   |  __| _` | |
                 | |   |    |   | (    |   |   | |   (   | |
           _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
Initializing KratosStructuralMechanicsApplication...
StructuralMechanicsAnalysis: Using the old way to create the processes, this will be removed!
Refer to "https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users#analysisstage-usage" for a description of the new format 
_/
           Multi-Physics 6.0.0-693bc1d484-FullDebug
Importing  KratosStructuralMechanicsApplication 
    KRATOS   ___|  |                   |                   |
           \___ \  __|  __| |   |  __| __| |   |  __| _` | |
                 | |   |    |   | (    |   |   | |   (   | |
           _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
Initializing KratosStructuralMechanicsApplication...
StructuralMechanicsAnalysis: Using the old way to create the processes, this will be removed!
Refer to "https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users#analysisstage-usage" for a description of the new format 
~~~

**Comment by philbucher** [comment]
your processes are not wrapped in a "processes" block
=> check out https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users#analysisstage-usage gor how to use it (it is very simple to fix)

**Comment by RiccardoRossi** [comment]
actually i have no processes at all...

**Comment by RiccardoRossi** [comment]
I should have written: i deliberately have no processes (everything is done automatically)

**Comment by philbucher** [comment]
ok now I understand

Fix incoming ...

File: applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
+# makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py]: this file should be names `rve_analysis.py`

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+from __future__ import print_function, absolute_import, division
+import KratosMultiphysics
+import KratosMultiphysics.StructuralMechanicsApplication as StructuralMechanicsApplication
+import KratosMultiphysics.KratosUnittest as KratosUnittest
+import KratosMultiphysics.kratos_utilities as kratos_utilities
+import RVEAnalysis
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]:
```suggestion
from KratosMultiphysics.StructuralMechanicsApplication.rve_analysis import RveAnalysis
```

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        Canalytic[3, 3] = G
+        Canalytic[4, 4] = G
+        Canalytic[5, 5] = G
+
+        for i in range(0, Cestimated.Size1()):
+            for j in range(0, Cestimated.Size2()):
+                self.assertAlmostEqual(
+                    abs(Cestimated[i, j] - Canalytic[i, j])/(l+2*G), 0.0, 5)
+
+        if not parameters["rve_settings"]["print_rve_post"].GetBool():
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:106]: this check is not strictly needed, would also work without

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+        parameters["solver_settings"]["block_builder"].SetBool(False)
+        parameters["solver_settings"]["multi_point_constraints_used"].SetBool(True)
+
+        self._aux_rve_computation(parameters)
+
+    def _aux_rve_computation(self, parameters):
+
+        model = KratosMultiphysics.Model()
+        simulation = RVEAnalysis.RVEAnalysis(model, parameters)
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]:
```suggestion
        simulation = RveAnalysis(model, parameters)
```

File: applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+from __future__ import print_function, absolute_import, division
+
+import KratosMultiphysics
+import KratosMultiphysics.StructuralMechanicsApplication
+import math
+from KratosMultiphysics.StructuralMechanicsApplication.structural_mechanics_analysis import StructuralMechanicsAnalysis
+
+
+class RVEAnalysis(StructuralMechanicsAnalysis):
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py]:
```suggestion
class RveAnalysis(StructuralMechanicsAnalysis):
```

**Review by philbucher — COMMENTED** [review]
some comments related to naming and importing

File: applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py (hunk 3 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py]: ok

File: applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py (hunk 7 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py]: I would prefer keeping the capitals here, as it is 

Represenative Volume Element

Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py]: ok, no problem

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+import KratosMultiphysics.StructuralMechanicsApplication as StructuralMechanicsApplication
+import KratosMultiphysics.KratosUnittest as KratosUnittest
+
+import KratosMultiphysics.kratos_utilities as kratos_utilities
+
+import RVEAnalysis
+
+import os
+
+class controlledExecutionScope:
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]: please use the new class in `kratos_utilities` for this

**Review by philbucher — CHANGES_REQUESTED** [review]
some things need to be cleaned up (my comments and the changes to the B&S)

**Comment by RiccardoRossi** [comment]
ok. Anyway this goes in after release

**Comment by loumalouomega** [comment]
> ok. Anyway this goes in after release

@RiccardoRossi can you approve the PR for my B&S?

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 5 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]: The point is that if i want the output i need the test not to delete the outputs...

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 4 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]: done (still calling it RVEAnalysis instead of RveAnalysis)

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 5 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:106]: i leave the check since i wish to be able to actually see the output if i set the print_rve_post to true

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 8 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]: @philbucher i am not clear what you are asking for. could u point me to an example?

Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]: https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/tests/structural_mechanics_test_factory.py#L13

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 4 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]: fine for me, as you explained me why :+1:

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 5 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:106]: ok I see

I still have a suggestion for you: From what I understand you always print the output, even if you don't want to see it (in 99% of the cases this test is run)
=> this is time consuming
=> I recommend to do it the other way round, explicitly only adding the printing of output if you want it, take a look what I am doing [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MeshMovingApplication/tests/test_ale_fluid_solver.py#L48-L101). This way the test won't get slowed down for nothing

**Comment by philbucher** [comment]
@RiccardoRossi if you are done with the changes I can also do some minor cleanup myself, this would be more efficient than having to iterate here :)

Just give me a go and I will do it according to the latest standards

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 8 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py]: thx, i think i corrected this

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 5 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:106]: It is only been printed if the parameters["rve_settings"]["print_rve_post"] is True

**Comment by RiccardoRossi** [comment]
@philbucher ok thx.

**Comment by RiccardoRossi** [comment]
mmm @philbucher i mean, i just commiteed my latest change

File: applications/StructuralMechanicsApplication/tests/test_rve_analytic.py (hunk 5 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:106]: but why would you then need to delete the files at all? If you want to print them you would want to look at them right?

Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:106]: ok I got it now, fixing it

**Comment by philbucher** [comment]
I cleaned up a bit, but I noticed that there are some occasional errors (even without my modifications, thats why I haven't pushed them yet):
Can you take a look please?
~~~
======================================================================
FAIL: test_rve_computation_elimination_version (__main__.TestRVESimplestTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_rve_analytic.py", line 36, in test_rve_computation_elimination_version
    self._aux_rve_computation(parameters)
  File "test_rve_analytic.py", line 74, in _aux_rve_computation
    abs(Cestimated[i, j] - Canalytic[i, j])/(l+2*G), 0.0, 5)
AssertionError: nan != 0.0 within 5 places

----------------------------------------------------------------------
~~~

**Comment by RiccardoRossi** [comment]
mmm i tried with valgrind and it looks like there is still something wrong with the EliminationB&S

@loumalouomega i am afraid this is on you.If you prefer i can take care of it

Right now i commented the test with the elimination builder and solver and left the default (blockB&S)
@philbucher could u see if it works correctly for you?

**Comment by loumalouomega** [comment]
> mmm i tried with valgrind and it looks like there is still something wrong with the EliminationB&S
> 
> @loumalouomega i am afraid this is on you.If you prefer i can take care of it
> 
> Right now i commented the test with the elimination builder and solver and left the default (blockB&S)
> @philbucher could u see if it works correctly for you?

When you say elimination you mean the elimination with constraints?

**Comment by philbucher** [comment]
The problems I had were only with the elimination-based

**Comment by RiccardoRossi** [comment]
@loumalouomega yes. In any case i would still merge this branch if it works, since the fix is unrelated

**Comment by loumalouomega** [comment]
> @loumalouomega yes. In any case i would still merge this branch if it works, since the fix is unrelated

In my machine the B&S works. Which is the problem?

**Comment by loumalouomega** [comment]
Besides, the test is still generating the auxiliar txt. This should not happen

**Comment by RiccardoRossi** [comment]
@loumalouomega  the txt is not auxiliar, it is the output of the RVEanalysis

regarding the Elimination B&S just try to compile this branch with FullDebug & -D_GLIBCXX_DEBUG and run it with valgrind

**Comment by philbucher** [comment]
@RiccardoRossi I will push the cleanup on Monday latest then this can go in, since the problems are in the B&S
Do you agree?

**Comment by loumalouomega** [comment]
> @loumalouomega the txt is not auxiliar, it is the output of the RVEanalysis
> 
> regarding the Elimination B&S just try to compile this branch with FullDebug & -D_GLIBCXX_DEBUG and run it with valgrind

Yes, already done. I created an independent PR on  #4400

**Comment by RiccardoRossi** [comment]
@philbucher ok thx

**Review by philbucher — DISMISSED** [review]

**Comment by philbucher** [comment]
I pushed some cleanup, hope its ok

Tests still running, but I sometimes still get the following error (FullDebug with two threads):
~~~
======================================================================
FAIL: test_rve_computation_elimination_version (__main__.TestRVESimplestTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_rve_analytic.py", line 33, in test_rve_computation_elimination_version
    self._aux_rve_computation(parameters)
  File "test_rve_analytic.py", line 103, in _aux_rve_computation
    abs(Cestimated[i, j] - Canalytic[i, j])/(l+2*G), 0.0, 5)
AssertionError: nan != 0.0 within 5 places
~~~

**Comment by loumalouomega** [comment]
> I pushed some cleanup, hope its ok
> 
> Tests still running, but I sometimes still get the following error (FullDebug with two threads):
> 
> ```
> ======================================================================
> FAIL: test_rve_computation_elimination_version (__main__.TestRVESimplestTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "test_rve_analytic.py", line 33, in test_rve_computation_elimination_version
>     self._aux_rve_computation(parameters)
>   File "test_rve_analytic.py", line 103, in _aux_rve_computation
>     abs(Cestimated[i, j] - Canalytic[i, j])/(l+2*G), 0.0, 5)
> AssertionError: nan != 0.0 within 5 places
> ```

Why it only fails in others computer?, well take into account I reverted the pointers due to @RiccardoRossi suggestion (and this was one of the reasons of valgrind messages)

**Comment by philbucher** [comment]
I had the latest modifications

**Comment by loumalouomega** [comment]
> I had the latest modifications

Yes, but I reverted the pointers change due to @RiccardoRossi suggestions (in any case I cannot duplicate the error)

**Comment by RiccardoRossi** [comment]
@loumalouomega but did u fix the error in some other way?

**Comment by loumalouomega** [comment]
> @loumalouomega but did u fix the error in some other way?

YEs, thta's why I did it. I assume that the error must be in some swap...but I cannot replicate it

**Review by loumalouomega — APPROVED** [review]
Approving and commenting test until fixed random fail on elimination B&S

**Comment by loumalouomega** [comment]
@RiccardoRossi feel free to merge
